### PR TITLE
Added "Merge roads" tool

### PR DESCRIPTION
**Pull request by Atria1234**
Added tool for finding and merging largest area of roads into single object. This will reduce the amount of placed objects on canvas. Possibly increasing performance, reducing save size.

### CODE REVIEW

File: AnnoDesigner/Helper/AdjecentCellGrouper.cs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        private (int y, int width, int height) FindLargestAreaUnderHistogram(int[] column)
+        {
+            var maxY = 0;
+            var maxWidth = 0;
+            var maxHeight = 0;
+
+            var stack = new Stack<int>(column.Length);
+            var y = 0;
+
+            /*void CalculateAreaAndUpdate()
```
Comment (FroggieFrog) [inline AnnoDesigner/Helper/AdjecentCellGrouper.cs:128]: Is there a reason this comment is still present?

Comment (Atria1234) [reply AnnoDesigner/Helper/AdjecentCellGrouper.cs]: I was testing the performance for possible use in irrigation coverage calculation, but it is a lost cause, so I'll revert it.

File: AnnoDesigner/ViewModels/MainViewModel.cs [hunk 2]
```diff
@@ -945,6 +946,46 @@ private void CanvasNormalize(object param)
...
+        /// <summary>
+        /// Filters all roads in current layout, finds largest groups of them and replaces them with merged variants.
+        /// </summary>
+        private void MergeRoads(object param)
+        {
+            var minX = AnnoCanvas.PlacedObjects.Where(p => p.WrappedAnnoObject.Road).Min(p => (int)p.Position.X);
+            var maxX = AnnoCanvas.PlacedObjects.Where(p => p.WrappedAnnoObject.Road).Max(p => (int)(p.Position.X + p.Size.Width));
+            var minY = AnnoCanvas.PlacedObjects.Where(p => p.WrappedAnnoObject.Road).Min(p => (int)p.Position.Y);
+            var maxY = AnnoCanvas.PlacedObjects.Where(p => p.WrappedAnnoObject.Road).Max(p => (int)(p.Position.Y + p.Size.Height));
+            var cells = Enumerable.Range(0, maxX - minX).Select(i => new LayoutObject[maxY - minY]).ToArray();
```
Comment (AgmasGold) [inline AnnoDesigner/ViewModels/MainViewModel.cs:960]: Can't we use the `StatisticsCalculationHelper` for this?

File: AnnoDesigner/ViewModels/MainViewModel.cs [hunk 3]
```diff
@@ -945,6 +946,46 @@ private void CanvasNormalize(object param)
...
+            foreach (var item in AnnoCanvas.PlacedObjects.Where(p => p.WrappedAnnoObject.Road))
```
Comment (AgmasGold) [inline AnnoDesigner/ViewModels/MainViewModel.cs:961]: Could we have braces around this for readability please? At least around the outer foreach - I can kind of see not doing it for the inner for loops.

**Comment by FroggieFrog** [comment]
Maybe the algorithm can also respect the color of the road.
If I load the layout and apply the optimization, than this particular layout changes in a not so good way.

#### before

before

#### after

after

#### The layout

[Bigger_City_with_Palace_World_Fair_Centered.zip](https://github.com/AnnoDesigner/anno-designer/files/5565340/Bigger_City_with_Palace_World_Fair_Centered.zip)

File: Tests/AnnoDesigner.Tests/AdjecentCellGrouperTests.cs [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+            for (var i = 0; i < gridLines.Length; i++)
+                for (var j = 0; j < gridLines[i].Length; j++)
+                    postTranspose[j][i] = preTranspose[i][j];
+
+            return postTranspose;
+        }
+
+        [Fact]
+        public void WholeArray()
```
Comment (AgmasGold) [inline Tests/AnnoDesigner.Tests/AdjecentCellGrouperTests.cs:42]: Could we rename these tests so they are in line with rest of the tests in the codebase? The usual pattern we're following is <method name>_<test conditions>_<expected result>, just look at some other tests for reference. The sections can be optional if it doesn't make sense - its not a hard and fast rule.

Basically, you want to be able to read the test from the test runner and know what its doing. If I read `WholeArray`, that doesn't explain anything to me. If I read `GroupAdjacentCells_SquareArrayNoGaps_ReturnsSingleResult`, then I know exactly what the test is actually doing.

**Review by AgmasGold — CHANGES_REQUESTED** [review]
Looks amazing, just added a couple of things.

**Comment by Atria1234** [comment]
> Maybe the algorithm can also respect the color of the road.
> If I load the layout and apply the optimization, than this particular layout changes in a not so good way.

Can you think of any other criteria? Don't merge roads without border with roads with border?

**Comment by FroggieFrog** [comment]
> Can you think of any other criteria? Don't merge roads without border with roads with border?

Personally I have no layout which uses borderless roads and roads with borders.
But it is possible that some users have this and therefor I think the algorithm should respect that too.
In general these are the onky 2 options for a road (color and border) so the amount of extra work should not be that much.

File: Tests/AnnoDesigner.Tests/AdjecentCellGrouperTests.cs [hunk 5]
```diff
@@ -58,7 +58,7 @@ public void WholeArray()
         }
 
         [Fact]
-        public void Chestboard()
+        public void MergeItems_ChestboardWithoutSingleCellMerge_NothingFound()
```
Comment (AgmasGold) [inline Tests/AnnoDesigner.Tests/AdjecentCellGrouperTests.cs:61]: This should be `Chessboard` 😄. Good job with the naming though!

**Review by AgmasGold — COMMENTED** [review]
Might also be worth adding some tests to ensure the border and background colours are not changed.

File: AnnoDesigner/Models/IAdjecentCellGrouper.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
+﻿using System.Collections.Generic;
+using System.Windows;
+
+namespace AnnoDesigner.Models
+{
+    public class CellGroup<T>
```
Comment (FroggieFrog) [inline AnnoDesigner/Models/IAdjecentCellGrouper.cs:6]: As far as I know the best practice in .Net is to put every class/interface in its own file.
So maybe move `CellGroup<T>` to its own file?
Also it is no private class, so one point more for its own file.

**Comment by FroggieFrog** [comment]
I found no issues using the functionality in this PR when it was used in some of my layout files.
@AgmasGold Do you have some issues/comments left open?

**Pull request merged**